---
title: "SpaTalk tutorial (single-cell ST data)"
author: "Xin Shao"
date: "`r Sys.Date()`"
output:
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## **Preprocess**

Before create SpaTalk object, we suggest to revise gene symbols of ST data `matrix` and scRNA-seq data `matrix` using `rev_gene()`

```{r load st data, echo=TRUE}
library(SpaTalk)

# load starmap data
load(paste0(system.file(package = 'SpaTalk'), "/extdata/starmap_data.rda"))
load(paste0(system.file(package = 'SpaTalk'), "/extdata/starmap_meta.rda"))

# revise gene symbols according to the NCNI gene symbols
starmap_data <- rev_gene(data = as.matrix(starmap_data),
                         data_type = "count",
                         species = "Mouse",
                         geneinfo = geneinfo)

```


## **Note**

In order to guide users on how to create the right data type, we provide several `demo_data` to follow:

```{r demo data, echo=TRUE}
# demo spot-based ST data
demo_st_data()

# demo spot-based ST meta
demo_st_meta()

# demo single-cell ST data
demo_st_sc_data()

# demo single-cell ST meta
demo_st_sc_meta()

# demo scRNA-seq data
demo_sc_data()

# demo geneinfo
demo_geneinfo()

```

After revising gene symbols, we create a SpaTalk object with st_data and st_meta. Given the two types of ST data, namely single-cell and spot-based ST data, we provided two standard processing examples. Please refer to:

## [**`single-cell tutorial vignette`**](https://raw.githack.com/ZJUFanLab/SpaTalk/main/vignettes/sc_tutorial.pdf)

## [**`spot-based tutorial vignette`**](https://raw.githack.com/ZJUFanLab/SpaTalk/main/vignettes/spot_tutorial.pdf)

```{r sessionInfo, echo=TRUE}
sessionInfo()
```



