---
title: "SpaTalk tutorial (spot-based ST data)"
author: "Xin Shao"
date: "`r Sys.Date()`"
output:
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
vignette: >
  %\VignetteIndexEntry{SpaTalk tutorial (spot-based ST data)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## **Main**

After pre-processing, we create a SpaTalk object with st_data and st_meta for spot-based ST data. In practise, we generate simulated spot-based ST from single-cell STARmap data using `generate_spot()`

```{r load st data, echo=TRUE}
library(SpaTalk)

# load starmap data
load(paste0(system.file(package = 'SpaTalk'), "/extdata/starmap_data.rda"))
load(paste0(system.file(package = 'SpaTalk'), "/extdata/starmap_meta.rda"))

# generate spot-based ST from single-cell STARmap data
st_data <- generate_spot(st_data = as.matrix(starmap_data),
                         st_meta = starmap_meta,
                         x_min = 0,
                         x_res = 200,
                         x_max = 3800,
                         y_min = 0,
                         y_res = 500,
                         y_max = 17500)
st_meta <- st_data$st_meta
st_data <- st_data$st_data

# create SpaTalk data
obj <- createSpaTalk(st_data = as.matrix(st_data),
                     st_meta = st_meta[, c(1,2,3)],
                     species = "Mouse",
                     if_st_is_sc = F,
                     spot_max_cell = max(st_meta$cell_real))

```

### Use `plot_st_pie_generate()` to view the cell-type composition

```{r plot_st_pie_generate, echo=TRUE}
plot_st_pie_generate(st_meta = st_meta, pie_scale = 1.3, xy_ratio = 1.8)

```

## **Cell-type decomposition**

First, we apply [NNLM](https://github.com/linxihui/NNLM) and spatial mapping to reconstruct the ST data at single-cell resolution using `dec_celltype()`

```{r cell-type decomposition, echo=TRUE}
# decode the cell-type composition
obj <- dec_celltype(object = obj,
                    sc_data = as.matrix(starmap_data),
                    sc_celltype = starmap_meta$celltype)

```

### Use `plot_st_pie()` to view predicted cell-type composition

```{r plot_st_pie, echo=TRUE}
# Scatter pie plot for each spot
plot_st_pie(object = obj, pie_scale = 1.3, xy_ratio = 1.8)

```

### Use `plot_st_celltype_percent()` to view cell-type percent

```{r plot_st_celltype_percent, echo=TRUE}
# plot cell-type percent across spatial spots
plot_st_celltype_percent(object = obj, celltype = 'Oligo',size = 3)

```

### Use `plot_st_gene()` to view gene expression

```{r plot_st_gene, echo=TRUE}
# plot marker gene expression across spatial spots
plot_st_gene(object = obj, gene = 'Plp1',size = 3, if_use_newmeta = F)

```

### Use `plot_st_cor_heatmap()` to view correlation heatmap

```{r plot_st_cor_heatmap, echo=TRUE}
# correlation between marker gene expression and cell type percent across spatial spots
plot_st_cor_heatmap(object = obj,
                    marker_genes = c("Plp1","Vip","Sst","Lamp5","Pcp4","Mfge8","Pvalb"),
                    celltypes = c("Oligo","VIP","SST","eL2_3","eL6","Astro","PVALB"),
                    scale = "none",
                    color_low = 'blue',
                    color_high = 'yellow',
                    color_mid = 'yellow')
```

## **ST at single-cell resolution**
### Use `plot_st_celltype()` to view cell type in space

```{r plot_st_celltype, echo=TRUE}
# plot cell type in reconstructed ST atlas
plot_st_celltype(object = obj, celltype = 'Oligo')

```

### Use `plot_st_gene()` to view gene expression in space

```{r plot_st_gene2, echo=TRUE}
# plot marker gene expression in reconstructed ST atlas
plot_st_gene(object = obj,gene = 'Plp1', if_use_newmeta = T)

```

### Use `plot_st_celltype_density()` to view cell-type density in space

```{r plot_st_celltype_density, echo=TRUE}
# plot cell-type density in reconstructed ST atlas
plot_st_celltype_density(object = obj,
                         celltype = 'Oligo',
                         type = 'raster',
                         color_low = 'purple',
                         color_high = 'yellow')

plot_st_celltype_density(object = obj,
                         celltype = 'Oligo',
                         type = 'contour',
                         color_low = 'purple',
                         color_high = 'yellow')
```

### Use `plot_st_celltype_all()` to view all cell types in space

```{r plot_st_celltype_all, echo=TRUE}
# plot all cell types in reconstructed ST atlas
plot_st_celltype_all(object = obj)

```

## **Infer cell-cell communications**
Based on the constructed ST atlas at single-cell resolution, we use `find_lr_path()` and `dec_cci()` to infer cell-cell communications mediated by ligand-receptor interactions (LRIs) in space using [LRIs from CellTalkDB](http://tcm.zju.edu.cn/celltalkdb/), pathways from [KEGG](https://www.genome.jp/kegg/pathway.html) and [Reactome](https://reactome.org/), and transcriptional factors (TFs) from [AnimalTFDB](http://bioinfo.life.hust.edu.cn/AnimalTFDB/#!/)

```{r dec_cci, echo=TRUE}
# Filter LRIs with downstream targets
obj <- find_lr_path(object = obj, lrpairs = lrpairs, pathways = pathways)

# Infer cell-cell communications from SST to PVALB neurons
obj <- dec_cci(object = obj, 
               celltype_sender = 'SST',
               celltype_receiver = 'PVALB')

# Get LR and downstream pathways
obj_lr_path <- get_lr_path(object = obj,
                           celltype_sender = 'SST',
                           celltype_receiver = 'PVALB',
                           ligand = 'Sst',
                           receptor = 'Sstr2')

obj_lr_path$tf_path
obj_lr_path$path_pvalue
```

### Use `plot_ccdist()` to view distribution of senders and receivers

```{r plot_ccdist, echo=TRUE}
# Point plot with spatial distribution of celltype_sender and celltype_receiver
plot_ccdist(object = obj, celltype_sender = 'SST', celltype_receiver = 'PVALB')

```

### Use `plot_cci_lrpairs()` to view all LRIs of senders and receivers

Given the limited LR pairs, we do not show the result here

```{r plot_cci_lrpairs, echo=TRUE}
# Heatmap with LR pairs of celltype_sender and celltype_receiver
# plot_cci_lrpairs(object = obj, celltype_sender = 'SST', celltype_receiver = 'PVALB')

```

### Use `plot_lrpair()` to view the specific LRI

```{r plot_lrpair, echo=TRUE}
# Point plot with LR pair from celltype_sender to celltype_receiver
plot_lrpair(object = obj,
            celltype_sender = 'SST',
            ligand = 'Sst',
            celltype_receiver = 'PVALB',
            receptor = 'Sstr2',
            if_plot_density = F)

```

### Use `plot_lrpair_vln()` to view violin plot with spatial distance of LRI

```{r plot_lrpair_vln, echo=TRUE}
# Violin plot with spatial distance of LR pair between senders and receivers and between all cell-cell pairs
plot_lrpair_vln(object = obj,
                celltype_sender = 'SST',
                ligand = 'Sst',
                celltype_receiver = 'PVALB',
                receptor = 'Sstr2')

```

### Use `plot_lr_path()` to view network with LR and downstream pathways

```{r plot_lr_path, echo=TRUE}
# Plot network with LR and downstream pathways
plot_lr_path(object = obj,                
             celltype_sender = 'SST',
             ligand = 'Sst',
             celltype_receiver = 'PVALB',
             receptor = 'Sstr2')

```

### Use `plot_path2gene()` to view river plot of significantly activated pathways

```{r plot_path2gene, echo=TRUE}
# River plot of significantly activated pathways and related downstream genes of receptors
plot_path2gene(object = obj,                
               celltype_sender = 'SST',
               ligand = 'Sst',
               celltype_receiver = 'PVALB',
               receptor = 'Sstr2')

```

## **Note**

To infer all paired cell-cell communications, use `dec_cci_all()` instead of `dec_cci()`

```{r dec_cci_all, echo=TRUE}
# Infer cell-cell communications from SST to PVALB neurons
# obj <- dec_cci_all(object = obj)
```

```{r sessionInfo, echo=TRUE}
sessionInfo()
```



